ssl/ssl_sess.c

```diff
@@ -437,12 +449,12 @@ int ssl_get_new_session(SSL *s, int session)
         }
 
         /* Choose which callback will set the session ID */
-        CRYPTO_r_lock(CRYPTO_LOCK_SSL_CTX);
+        CRYPTO_THREAD_read_lock(s->lock);
         if (s->generate_session_id)
             cb = s->generate_session_id;
         else if (s->session_ctx->generate_session_id)
```

Comment:
There is a lock -mess between SSL * s and SSL_CTX * and SSL_CTX *
The variable  's ->generate_session_id'  and 's->session_ctx->generate_session_id' are write using two distinct  locks, see upper...
 It's unrelated to this conversion work.